### ISSUE TITLE
Windows: Verify TestExplorer works

### ISSUE THREAD

**Issue by adam-fowler** [swift-server/vscode-swift#188]
The next release of vscode-swift will include the new TestExplorer. We need to verify this works on Windows. Test discovery requires a built project as it uses `swift test --list-tests --skip-build` to get a list of the tests. Once we have a list of the tests the LSP server is used whenever a file is saved to add/remove tests to `XCTestCase` classes that have already been registered. You can find the TestExplorer by selecting the chemical container icon on the left hand side of the tree view container.

Before starting tests open a project which includes tests in VSCode and select the command `Swift: Clean Build` and then restart VSCode.

Things to test

- [ ] As we are starting with a clean build the test explorer should show a `Test Discovery Error` which when you open tells you to build the project.
- [ ] Build your project `Ctrl + Shift + B`. Once the build has completed the TestExplorer view should update to list the tests in your project. You should have a tree view item for each test target, inside the test target item should be a list of the `XCTestCase` classes and then inside them should be a list of the test methods. (Pretty much the same as Xcode)
- [ ] Select a single test. You should the option to run the test, or run the test in the debugger. If you select either of these it will build your project and then run your test. If you select `Run Test` then break points will not work. If you select `Debug Test` break points will work. I know the debugger is a little flakey so I'm not sure how much you can test debugging of tests.
- [ ] If you select run/debug tests on a class it should run all the tests in that class
- [ ] If you select run/debug tests on a target it should run all the tests in that target.
- [ ] Open a file for a test target and add a new test method. When you save the file it should add the new test to the TestExplorer.
- [ ] Open a file for a test target and remove a test method. When you save the file it should remove that test from the TestExplorer.

**Comment by compnerd** [comment]
Although the test enumeration works and in theory the test execution works, this is slightly broken for an important test case.  We don't honor adjustments of the build arguments when building the test.  If I have customized the flags, the new flags are discarded when a test is actually executed (resulting in both a rebuild and a failure to build).

**Comment by adam-fowler** [comment]
Are you sure this is the case? I use exactly the same build task when running the tests as when doing a simple build (triggered by Shift+Ctrl+B). They both use the build arguments that are in the extension configuration. 

I'll verify this. Although I can only verify on macOS or Linux. 

Out of interest what build arguments were you using.

**Comment by adam-fowler** [comment]
Just verified on macOS and Linux via a dev container. If I set the `swift: Build Arguments` setting those build arguments are used when building before running a test as well as building via `Run Build Task...`.

**Comment by compnerd** [comment]
Trying to pass along `-Xcc -IS:\Library\sqlite-3.38.0\usr\include -Xlinker -LS:\Library\sqlite-3.38.0\usr\lib -Xlinker C:\Library\Developer\Platforms\Windows.platform\Developer\SDKs\Windows.sdk\usr\lib\swift\windows\x86_64\swiftCore.lib`.

Sounds like I must have configured the settings incorrectly?  The build actually used it.  To be fair, I just saw the gear icon, clicked it and added in the options.  Do they need to be added in some other way?  It could be applied for a specific build.

**Comment by adam-fowler** [comment]
Did you add each option in separately? I've found it doesn't like it if you add them all in one line. You need to add them individually
```
-Xcc
-IS:\Library\sqlite-3.38.0\usr\include
-Xlinker
-LS:\Library\sqlite-3.38.0\usr\lib
etc
```

**Comment by adam-fowler** [comment]
If you can build the extension, you could comment out line 90 in SwiftTaskProvider.ts
https://github.com/swift-server/vscode-swift/blob/f8da2d740990ecdb5b4803a55b6eeee20253cd71/src/SwiftTaskProvider.ts#L83-L92
This will not clear the console window when it runs a new task and you'll be able to see the command line for the task.

**Comment by compnerd** [comment]
Yes, they are individual, with quoting of the argument and the backslashes (as it is an escape sequence otherwise).

The entry is marked as:
```
"group": "build",
"label": "swift: Build All",
"detail": "swift build --build-tests -Xswiftc -g -Xswiftc -use-ld=lld -Xlinker -debug:dwarf"
```

I'll build the extension in a bit

**Comment by compnerd** [comment]
BTW, perhaps it would make sense to have the flag be controllable by the user for debugging purposes in the future?

**Comment by adam-fowler** [comment]
Ideally VSCode wouldn't delete the `starting task` text when it is set to true. The purpose of the clear flag is to clear results from previous task runs , not delete the `starting task` text.

I can still add an option to enable/disable it. I guess some people might like to see results from old builds

**Comment by compnerd** [comment]
Ah; the starting task definitely doesn't show the flags.  And I realized that I didn't mention what I am building (which would be useful for someone else trying to replicate this) - it is just swift-driver.

**Comment by adam-fowler** [comment]
When you run the build task via `Run Build Task...` do you get the flags?

**Comment by compnerd** [comment]
Correct; the build task does get the flags (so Ctrl+Shift+B will succeed).  When you then run the tests, it will go, oh flags are different, rebuild and fail to build.

**Comment by adam-fowler** [comment]
Wait a sec, have you created a separate task in `tasks.json` to add in your extra command line params? That might be the cause, I rebuild the build task before running it in the case of building tests and will use the build parameters set in the settings. I should be able to fix it so I use the edited task from tasks.json if it is there.

**Comment by compnerd** [comment]
I had just clicked on the gear icon that showed up when I first went to build the project.  I believe that it duplicated the task to add the flags.  Is there another way to apply flags?  If so, I think that exposing that might be a better approach.

**Comment by adam-fowler** [comment]
I should be able to support both. 

Currently though the method is
- Go to Settings (Ctrl+,)
- Expand `Extensions` section
- Select `Swift`
- Second in the list is a `Swift: Build Arguments`. You can add all your build arguments there.

**Comment by compnerd** [comment]
Ah, interesting.  However, those options would be globally applied right?  I think that we need a way to support per-project flags.  The reasoning for that is simply that system libraries on Windows will likely need to be configured by the user (pkg-config is not common on Windows, and the user may install libraries wherever they like).

**Comment by adam-fowler** [comment]
You can set them at a workspace level (see at the top there is a User/Workspace selection)

**Comment by compnerd** [comment]
Ah ha!  Okay, so that is simply a PEBKAC.  Sorry about the long discussion on a completely unrelated item.  I guess that it uncovered something that someone unfamiliar with VSCode might trip over.

**Comment by adam-fowler** [comment]
Oh no, this is great. We need to fix usability issues like this. 
Also shows where documentation might help.

**Comment by compnerd** [comment]
So, just to add a bit more color - the last time I used an IDE was back ~2005 when I was at Microsoft, and naturally that was Visual Studio.  As a result, I expected project build settings and IDE settings to be separate.  In my mind, the settings dialog that you directed me towards is related to the IDE.  It never occurred to me that there might be project settings there.  How difficult would it be to augment the UI to include a discoverability alteration that allows you to use a context menu or button to go to that location?  Although, I would understand if that would be deemed far outside the scope of this project.

**Comment by compnerd** [comment]
Test enumeration seems to work; hovering on a test gives the option to run a test or run under a debugger.  Go to test also seems to function properly - though seemed to open a new pane unnecessarily.  Execution seems to be an issue  though - while it runs the test as per the output, it does not provide the test output nor does it provide test results.

**Comment by adam-fowler** [comment]
> So, just to add a bit more color - the last time I used an IDE was back ~2005 when I was at Microsoft, and naturally that was Visual Studio. 

Ha ha I remember swearing at Visual Studio quite a bit back then, when I was in the games industry. It didn't like AAA title size projects much.

**Comment by adam-fowler** [comment]
> Test enumeration seems to work; hovering on a test gives the option to run a test or run under a debugger. Go to test also seems to function properly - though seemed to open a new pane unnecessarily. Execution seems to be an issue though - while it runs the test as per the output, it does not provide the test output nor does it provide test results.

Damn! are you getting the same results with both running tests and debugging tests? They are pushed through quite different paths. When running tests I run the test executable with `child_process.execFile`. I can then pipe stdout/stderr to a WriteStream which outputs to the test console, parses the output and then updates tests accordingly. When debugging tests I have a lot less control over stdout/stderr but I can pipe the output to a file which I parse once the debug session has completed. If tests take a long time it can seem the debugging tests is doing nothing.

**Comment by compnerd** [comment]
I haven't tested the debugging of the tests yet, but I can try that.  I'll run through `npm run test` later.

**Comment by adam-fowler** [comment]
Ok. From the `npm test` run I have ascertained two of the possible issues are not issues which is good.

Yes if the tests aren't built you need to build them for discovery. The Test Discovery Error should read "Build the project to enable test discovery" though.

There is a Swift output window (Select View->Output and then select Swift in the dropdown menu at the RHS of the view) which gives general information about what the extension is doing. If you could cut and paste this in here for the case where the project hasn't been built and displays a Test Discovery error and also in the case where you try to debug tests that'd be great. I'm going to add a diagnostics option to add a lot more input into this view, to help debugging this issue.

**Comment by adam-fowler** [comment]
I've added the extra diagnostics in the output view. If you could run the latest code from git, enable `Swift: Diagnostics` in the Settings and reload the window (found in command view Ctrl+Shift+P), or restart vscode. Then run the test debug again and after include the output from the swift output window here. 

Sorry for all this. I imagine you've got plenty of other stuff to be getting on with.

**Comment by compnerd** [comment]
```
18:06:22: compnerd.org Swift version 5.7-dev (LLVM 0c27db54bc46582, Swift c056aeb1bf0fdf7)
18:06:27: SourceKit-LSP setup for swift-driver
18:06:27: swift-driver: add: s:\SourceCache\swift-driver
18:06:27: swift-driver: Resolving Dependencies ... 18:06:36: swift-driver: resolvedUpdated: s:\SourceCache\swift-driver
18:06:40: swift-driver: resolvedUpdated: s:\SourceCache\swift-driver
done.
18:06:41: swift-driver: focus: s:\SourceCache\swift-driver
18:06:42: swift-driver: resolvedUpdated: s:\SourceCache\swift-driver
18:06:42: swift-driver: Resolving Dependencies ... 18:06:44: swift-driver: resolvedUpdated: s:\SourceCache\swift-driver
18:06:44: swift-driver: Test Discovery Failed: error: The operation could not be completed. The file doesn’t exist.

done.
18:08:47: swift-driver: resolvedUpdated: s:\SourceCache\swift-driver
18:08:47: swift-driver: Resolving Dependencies ... 18:08:48: swift-driver: resolvedUpdated: s:\SourceCache\swift-driver
18:08:48: swift-driver: Test Discovery Failed: error: The operation could not be completed. The file doesn’t exist.

done.
```

**Comment by compnerd** [comment]
```
18:13:23: compnerd.org Swift version 5.7-dev (LLVM 0c27db54bc46582, Swift c056aeb1bf0fdf7)
18:13:23: Toolchain Path: C:\Library\Developer\Toolchains\unknown-Asserts-development.xctoolchain
18:13:23: Developer Dir: C:\Library\Developer
18:13:23: XCTestPath: C:\Library\Developer\Platforms\Windows.platform\Developer\Library\XCTest-development\usr\bin
18:13:27: swift-driver: add: s:\SourceCache\swift-driver
18:13:27: swift-driver: Resolving Dependencies ... 18:13:37: swift-driver: resolvedUpdated: s:\SourceCache\swift-driver
done.
18:13:38: swift-driver: Exec: swift test --skip-build --list-tests
18:13:39: SourceKit-LSP setup for swift-driver
18:13:39: swift-driver: focus: s:\SourceCache\swift-driver
18:13:45: swift-driver: Exec: swift test --skip-build --list-tests
18:13:49: swift-driver: resolvedUpdated: s:\SourceCache\swift-driver
18:13:49: swift-driver: Resolving Dependencies ... 18:13:49: swift-driver: resolvedUpdated: s:\SourceCache\swift-driver
18:13:50: swift-driver: resolvedUpdated: s:\SourceCache\swift-driver
18:13:51: swift-driver: resolvedUpdated: s:\SourceCache\swift-driver
18:13:51: swift-driver: Test Discovery Failed: error: The operation could not be completed. The file doesn’t exist.

18:13:51: swift-driver: Test Discovery Failed: error: The operation could not be completed. The file doesn’t exist.

done.
```

Hmm, one thing that stands out as being a problem is the `XCTestPath`.  The paths for the XCTest runtime has changed in the recent times, and we need a backwards/forwards compatible change.  It used to be in `bin`, now it is architecture dependent:

| Target | Relative path |
| :--: | :--: |
| i686-unknown-windows-msvc | bin32 |
| x86_64-unknown-windows-msvc | bin64 |
| armv7-unknown-windows-msvc | bin32a |
| aarch64-unknown-windows-msvc | bin64a |

**Comment by compnerd** [comment]
Okay, digging through the test explorer code to a certain extent, it appears that it runs `swift test --skip-build --list-tests`.  However, running that from the command line yields: `error: The operation could not be completed. The file doesn’t exist.`.  Sounds like a bug in swift-package-manager.  Back over to me on the other side.  In the mean time, I've put up a couple of cleanups, and the bin directory handling needs to be taken care of still.

**Comment by compnerd** [comment]
No wait, that is a bug in the project.  Patch incoming.

![image](https://user-images.githubusercontent.com/63311/163094750-5d366ed0-dd0a-4c3d-8081-3803eb916940.png)

**Comment by compnerd** [comment]
Okay, with pending changes, and a change to the XCTestPath, debug test seems to work!
![image](https://user-images.githubusercontent.com/63311/163099162-47eff04c-b230-467f-b34b-32d2ea979d97.png)

The rebuild everything to run each individual test is really rather annoying, is that something with the VSCode plugin or a bug in s-p-m/s-d?  I know that from the command line, I normally do not have to rebuild everything.  Although, I am seeing weird errors in my local swift-driver test suite, so it _could_ be related to the recent swift dynamic cast change.  What about the name undecoration in the stack trace?

It appears that the test output is still missing though when you run the test :-(.  Interestingly enough, it also does not update the test results when run, but does when debugged?

**Comment by adam-fowler** [comment]
The rebuild is an SPM thing. I call `swift build` before running tests but if a project is built it should just return immediately.

The undecorated name, indicates that lldb hasn't been setup to use the swift lldb. Check the settings for CodeLLDB -> Advanced -> Library. That should be pointing to lldb dll is I remember correctly. The swift extension sets this up on initialisation.
 
At the moment when debugging tests the test output only appears once the tests have all finished. This is the case on all platforms. I haven't managed to get access to the stdout stream. 

When just running (not debugging) tests the stdout is piped to a stream which parses the data. On macOS and Linux this is done line by line. I'm wondering if on Windows this is coming in differently. If you stick a breakpoint in line 205 of TestRunner.ts and run tests and check what is coming through that would be useful to know.

**Comment by compnerd** [comment]
Sorry, you are going to have to walk me through this.  I don't know how to setup everything to set the breakpoint and debug the plugin :-(.

**Comment by adam-fowler** [comment]
> Sorry, you are going to have to walk me through this.  I don't know how to setup everything to set the breakpoint and debug the plugin :-(.

It should be as simple as
- Open up VSCode inside the vscode-swift folder
- Press F5, which should build and run the extension inside a new VSCode window.
- go back to original VSCode
- open up source file and click on LHS of text editor on line where you want breakpoint.

**Comment by compnerd** [comment]
> The undecorated name, indicates that lldb hasn't been setup to use the swift lldb. Check the settings for CodeLLDB -> Advanced -> Library. That should be pointing to lldb dll if I remember correctly. The swift extension sets this up on initialisation.

Nope, that appears to be empty.

**Comment by compnerd** [comment]
Okay, so #248 corrects the reporting - tests now show the status properly. They were silently failing previously.  They do not provide any output still.  I did manage to get the debugger attached, but received only a single callback with the first 3 lines of output. I'm not sure why the rest of the output is not being relayed.

**Comment by compnerd** [comment]
> When just running (not debugging) tests the stdout is piped to a stream which parses the data. On macOS and Linux this is done line by line. I'm wondering if on Windows this is coming in differently. If you stick a breakpoint in line 204 of TestRunner.ts

Ah ha!  It is.  It ends up coming in as a single blob, but never gets passed to the writable stream.

**Comment by compnerd** [comment]
> The rebuild is an SPM thing. I call swift build before running tests but if a project is built it should just return immediately.

This is not a SPM bug - running from the terminal, it does immediately return saying that everything is up to date.

**Comment by compnerd** [comment]
![image](https://user-images.githubusercontent.com/63311/163315447-15b7714f-3dfc-435d-8da6-b25347d8a85e.png)

Not sure, but with some logging enabled seems that the class and individual tests now run properly and even provide the output in the console.

**Comment by adam-fowler** [comment]
I'm going to break out some of the stuff here into separate issues to simplify things

**Comment by adam-fowler** [comment]
> Not sure, but with some logging enabled seems that the class and individual tests now run properly and even provide the output in the console.

Not sure what you mean here. Did you add some additional logging?

**Comment by compnerd** [comment]
Okay, the debugging part is really weird for a user.  You need to set a breakpoint AND enter the test for debugging.  Entering the test for just debugging is pointless as it just runs it under lldb and exits.  Setting the breakpoint and running is pointless because it won't be under lldb.  Implicitly setting the breakpoint on the test being run for debugging would be way more obvious (is there an equivalent to `tb`?).  But seems that the functionality at least is working with a small change.  Loading the Swift version of liblldb crashes due to a Swift load failure.  This is something that _can_ be worked around, but is annoying: https://bugs.swift.org/browse/SR-13591?focusedCommentId=64462&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-64462.  An initial lldb script to load the runtime would workaround the issue.

**Comment by adam-fowler** [comment]
Run tests means running tests outside of the debugger. So breakpoints won't work.
Debugging tests run tests inside the debugger. I guess if you aren't setting a breakpoint it is kinda pointless, but it isn't always obvious where the breakpoint should be set. You may be running multiple tests, you want to it to stop at a specific point in the test you are debugging not at the beginning etc

**Comment by compnerd** [comment]
Agreed that it isn't always obvious where to set the breakpoint.  However, if running a single test, it would be nice to break on the test entry.  For a suite, perhaps on suite construction or entry?

**Comment by adam-fowler** [comment]
Closing as I think we are good here now.
